ethcore/src/client/chain_notify.rs

```diff
@@ -29,15 +31,76 @@ pub enum ChainMessageType {
 	SignedPrivateTransaction(Vec<u8>),
 }
 
+/// Route type to indicate whether it is enacted or retracted.
+#[derive(Clone)]
+pub enum ChainRouteType {
+	/// Enacted block
+	Enacted,
+	/// Retracted block
+	Retracted
+}
+
+/// A complete chain enacted retracted route.
+#[derive(Default, Clone)]
+pub struct ChainRoute(pub Vec<(H256, ChainRouteType)>);
+
+impl<'a> From<&'a [ImportRoute]> for ChainRoute {
+	fn from(import_results: &'a [ImportRoute]) -> ChainRoute {
+		ChainRoute(import_results.iter().flat_map(|route| {
+			route.retracted.iter().map(|h| (*h, ChainRouteType::Retracted))
+				.chain(route.enacted.iter().map(|h| (*h, ChainRouteType::Enacted)))
+		}).collect())
+	}
+}
+
+impl ChainRoute {
+	/// Gather all non-duplicate enacted and retracted blocks.
+	pub fn to_enacted_retracted(&self) -> (Vec<H256>, Vec<H256>) {
```

Comment:
It seems that we call it for every `ChainRoute` anyway and sometimes we call it multiple times. Maybe calculate that in the constructor as well and just return `(&[H256], &[H256])` here?